pkg/executor/aggregate/agg_hash_executor.go

```diff
@@ -292,58 +272,118 @@ func (e *HashAggExec) initForParallelExec(_ sessionctx.Context) {
+			partialResultMap:           make(aggfuncs.AggPartialResultMapper),
+			BInMap:                     0,
+			inputCh:                    e.partialOutputChs[i],
+			outputCh:                   e.finalOutputCh,
+			finalResultHolderCh:        make(chan *chunk.Chunk, 1),
+			rowBuffer:                  make([]types.Datum, 0, e.Schema().Len()),
+			mutableRow:                 chunk.MutRowFromTypes(exec.RetTypes(e)),
+			groupKeys:                  make([][]byte, 0, 8),
+			spillHelper:                e.spillHelper,
+			restoredAggResultMapperMem: 0,
 		}
 		// There is a bucket in the empty partialResultsMap.
-		e.memTracker.Consume(hack.DefBucketMemoryUsageForMapStrToSlice*(1<<w.BInMap) + setSize)
-		groupSet.SetTracker(e.memTracker)
+		e.memTracker.Consume(hack.DefBucketMemoryUsageForMapStrToSlice * (1 << e.finalWorkers[i].BInMap))
 		if e.stats != nil {
-			w.stats = &AggWorkerStat{}
-			e.stats.FinalStats = append(e.stats.FinalStats, w.stats)
+			e.finalWorkers[i].stats = &AggWorkerStat{}
+			e.stats.FinalStats = append(e.stats.FinalStats, e.finalWorkers[i].stats)
 		}
-		e.finalWorkers[i] = w
 		e.finalWorkers[i].finalResultHolderCh <- exec.NewFirstChunk(e)
 	}
+}
+
+func (e *HashAggExec) initForParallelExec(ctx sessionctx.Context) error {
+	sessionVars := e.Ctx().GetSessionVars()
+	partialConcurrency := sessionVars.HashAggPartialConcurrency()
+	finalConcurrency := sessionVars.HashAggFinalConcurrency()
+
+	if partialConcurrency == 0 || finalConcurrency == 0 {
+		return errors.New("partialConcurrency or finalConcurrency is 0")
+	}
+
+	e.IsChildReturnEmpty = true
+	e.finalOutputCh = make(chan *AfFinalResult, finalConcurrency+partialConcurrency+1)
+	e.inputCh = make(chan *HashAggInput, partialConcurrency)
+	e.finishCh = make(chan struct{}, 1)
+
+	e.partialInputChs = make([]chan *chunk.Chunk, partialConcurrency)
+	for i := range e.partialInputChs {
+		e.partialInputChs[i] = make(chan *chunk.Chunk, 1)
+	}
+	e.partialOutputChs = make([]chan *aggfuncs.AggPartialResultMapper, finalConcurrency)
+	for i := range e.partialOutputChs {
+		e.partialOutputChs[i] = make(chan *aggfuncs.AggPartialResultMapper, partialConcurrency)
+	}
+
+	e.inflightChunkSync = &sync.WaitGroup{}
+
+	isTrackerEnabled := e.Ctx().GetSessionVars().TrackAggregateMemoryUsage && variable.EnableTmpStorageOnOOM.Load()
+	isParallelHashAggSpillEnabled := e.Ctx().GetSessionVars().EnableConcurrentHashaggSpill
 
+	baseRetTypeNum := len(e.RetFieldTypes())
+
+	// Intermediate result for aggregate function also need to be spilled,
+	// so the number of spillChunkFieldTypes should be added 1.
+	spillChunkFieldTypes := make([]*types.FieldType, baseRetTypeNum+1)
+	for i := 0; i < baseRetTypeNum; i++ {
+		spillChunkFieldTypes[i] = types.NewFieldType(mysql.TypeVarString)
+	}
+	spillChunkFieldTypes[baseRetTypeNum] = types.NewFieldType(mysql.TypeString)
+	e.spillHelper = newSpillHelper(e.memTracker, e.PartialAggFuncs, func() *chunk.Chunk {
```

Comment:
define a new tracker for spillHelper and attach this new tracker to e.memTracker

Reply:
> define a new tracker for spillHelper and attach this new tracker to e.memTracker

Most of unconsume are called in spillHelper and most of consume are called in partial worker. I think it's better to use the same tracker as memory should be consumed and unconsumed by the same tracker.